### DISCUSSION TITLE
Colorless async

### DISCUSSION THREAD

**Discussion by CommonLoon102** [dotnet/csharplang#4698] [Language Ideas] [answered]
I'd like to call async methods asynchronously by default, without having to go through all this pain:
- Write `await` before the method call
- Change the return type of the caller method to wrap the return type into `async Task<>`
- Append `Async` at the end of the name of the caller method (I could skip this but that wouldn't be nice)
- Do the above two again for the method signature in the interface as well (because interfaces are everywhere since dependency injection is cool)
- Do all of the above for every function which calls the caller function, recursively, up to the top (async main, razor.cs file, etc.)

The compiler could do this at compile time instead of me, silently in the background, so I wouldn't have to go through all of this time consuming, brainless, manual work all the time.

In Server-Side Blazor apps it can happen that I want to do a database call, a HTTP request or JavaScript interop. Then this little game begins. Why is it needed? Why can't the compiler just do it instead of me? What would be the drawbacks or pitfalls?

**Comment by HaloFour** [comment]
> The compiler could do this at compile time instead of me, silently in the background, so I wouldn't have to go through all of this time consuming, brainless, manual work all the time.

What are you suggesting that the compiler do here?  Make every single method that is written automatically async?

Asynchrony comes with additional overhead and various considerations that should be kept in mind.  Every asynchronous boundary comes with the possibility of the execution being suspended and resumed on a completely different thread.  The team decided that it was best to ensure that the developer was well aware of these boundaries and was required to opt-in to where they occurred.  Without all of the "boilerplate" you mentioned that asynchronous code just becomes imperative blocking code again.

**Reply by bbarry** [reply]
Fundamentally it is possible to write code with a sync path and use Roslyn to generate the async versions with something like this: https://github.com/roji/AsyncRewriter (that probably needs to be updated some but I think it could be converted to a Roslyn source generator package reasonably). Eg you write:


    class MyClass
    {
        [RewriteAsync]
        public int Foo(Stream s)
        {
            var data = new byte[10];
            return s.Read(data, 0, 10);
        }
    }


and save the file and a generated file gets added:


    partial class MyClass
    {
        public async Task<int> FooAsync(Stream s, CancellationToken cancellationToken)
        {
            var data = new byte[10];
            return await s.ReadAsync(data, 0, 10, cancellationToken);
        }
    }
   
@roji did this for a while in https://github.com/npgsql/npgsql and can probably provide better insight into why it isn't being done anymore but I suspect it boils down to "doing async right is hard and often isn't going to just be the same as the sync code but adding `async`/`await` and calling a different method here or there."

**Reply by roji** [reply]
Wow, people are somehow still aware of this :rofl: 

AsyncRewriter was a cool project, but ultimately turned out to be quite brittle... I'm no longer fresh on the details, but ultimately having a single sync/async path (with async methods and a syncOrAsync flag that gets passed down the stack) proved performant enough that two separate codepaths with code generation didn't seem necessary any more.

**Reply by alrz** [reply]
>What are you suggesting that the compiler do here? Make every single method that is written automatically async?

The way I see it, it all comes down to making this kind of refactoring unnecessary: https://github.com/microsoft/vs-threading/issues/746

That is purely syntactic. It could very well be an op-in change to make await implicit.

**Comment by HaloFour** [comment, marked as answer]
Here's a discussion that happened on this subject from the runtime perspective recently:

https://github.com/dotnet/runtime/issues/50796

**Reply by CommonLoon102** [reply]
> Here's a discussion that happened on this subject from the runtime perspective recently:
> 
> [dotnet/runtime#50796](https://github.com/dotnet/runtime/issues/50796)

Thanks, that's what I was thinking about.
Basically opting out instead of opting in, but only that itself would be a breaking change what nobody would want.

For my simple problem, a Roslyn code-fix would be enough, to eliminate the manual work. Having the compiler/runtime do it and changing the whole platform is just too much in every aspect.

Meanwhile @bbarry brought up the Roslyn code-fix solution I wanted to ask about, maybe somebody out there has already started something, and yes.

I cannot mark two answers as answers, but actually both of them were needed to fully answer my question.

**Reply by bugproof** [reply]
If you're using Rider or ReSharper it will most of the time detect that you're in async context and propose a fix to convert the call to async/await.
